arangod/RocksDBEngine/RocksDBCollection.cpp

```diff
@@ -388,104 +362,35 @@ void RocksDBCollection::getPropertiesVPack(velocypack::Builder& result) const {
   TRI_ASSERT(result.isOpenObject());
 }
 
-/// @brief closes an open collection
-ErrorCode RocksDBCollection::close() {
-  RECURSIVE_READ_LOCKER(_indexesLock, _indexesLockWriteOwner);
-  for (auto it : _indexes) {
-    it->unload();
-  }
-  return TRI_ERROR_NO_ERROR;
-}
-
 /// return bounds for all documents
 RocksDBKeyBounds RocksDBCollection::bounds() const {
   return RocksDBKeyBounds::CollectionDocuments(objectId());
 }
 
-void RocksDBCollection::prepareIndexes(
-    arangodb::velocypack::Slice indexesSlice) {
-  TRI_ASSERT(indexesSlice.isArray());
-
-  auto& selector =
-      _logicalCollection.vocbase().server().getFeature<EngineSelectorFeature>();
-  auto& engine = selector.engine<RocksDBEngine>();
-  std::vector<std::shared_ptr<Index>> indexes;
-  {
-    // link creation needs read-lock too
-    RECURSIVE_READ_LOCKER(_indexesLock, _indexesLockWriteOwner);
-    if (indexesSlice.length() == 0 && _indexes.empty()) {
-      engine.indexFactory().fillSystemIndexes(_logicalCollection, indexes);
-    } else {
-      engine.indexFactory().prepareIndexes(_logicalCollection, indexesSlice,
-                                           indexes);
-    }
-  }
+void RocksDBCollection::prepareIndexes(velocypack::Slice indexesSlice) {
+  PhysicalCollection::prepareIndexes(indexesSlice);
 
-  RECURSIVE_WRITE_LOCKER(_indexesLock, _indexesLockWriteOwner);
-  TRI_ASSERT(_indexes.empty());
-  for (std::shared_ptr<Index>& idx : indexes) {
+  // update tick value and _primaryIndex member
+  for (auto& idx : getIndexes()) {
     TRI_ASSERT(idx != nullptr);
-    auto const id = idx->id();
-    for (auto const& it : _indexes) {
-      TRI_ASSERT(it != nullptr);
-      if (it->id() == id) {  // index is there twice
-        idx.reset();
-        break;
-      }
-    }
-
-    if (idx) {
-      TRI_UpdateTickServer(static_cast<TRI_voc_tick_t>(id.id()));
-      _indexes.emplace(idx);
-      if (idx->type() == Index::TRI_IDX_TYPE_PRIMARY_INDEX) {
-        TRI_ASSERT(idx->id().isPrimary());
-        _primaryIndex = static_cast<RocksDBPrimaryIndex*>(idx.get());
-      }
+    TRI_UpdateTickServer(static_cast<TRI_voc_tick_t>(idx->id().id()));
+    if (idx->type() == Index::TRI_IDX_TYPE_PRIMARY_INDEX) {
+      TRI_ASSERT(idx->id().isPrimary());
+      _primaryIndex = static_cast<RocksDBPrimaryIndex*>(idx.get());
```

Comment:
Before `_primaryIndex` was set under `_indexesLock` but IDK if it's a problem or not. As an alternative I'd propose changing `PhysicalCollection::prepareIndexes` so it returns a pointer to primary index.